---
title: "R Assignment Notebook"
output: html_notebook
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Ctrl+Shift+Enter*. 

```{r}
plot(cars)
```

Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Ctrl+Alt+I*.
```{r}

```


When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Ctrl+Shift+K* to preview the HTML file).


Inspecting fang_et_al_genotypes.txt

```{r}
getwd()
setwd("C:/Users/Devin/Documents/GradSchool/IowaState/EEOB_546X/homework")
getwd()
genotype_data <- read.table("fang_et_al_genotypes.txt",header=TRUE,stringsAsFactors = FALSE)
dim(genotype_data)
colnames(genotype_data[1:10])
row.names(genotype_data[1:10])
```


Inspecting snp_position.txt
```{r}
snp_position <- read.delim("snp_position.txt",header=TRUE, stringsAsFactors = FALSE)
dim(snp_position)
colnames(snp_position[1:10])
row.names(snp_position[1:10])
```

Extract maize data from genotype_data.
```{r}
maize_genotype <- genotype_data[(genotype_data$Group == "ZMMIL") | (genotype_data$Group =="ZMMMR")| (genotype_data$Group == "ZMMLR"), ] 
nrow(maize_genotype)
ncol(maize_genotype)
unique(maize_genotype$Group)
```
Extract Teosinte data from genotype_data.
```{r}
teosinte_genotype <- genotype_data[(genotype_data$Group == "ZMPBA") | (genotype_data$Group =="ZMPIL")| (genotype_data$Group == "ZMPJA"), ] 
nrow(teosinte_genotype)
ncol(teosinte_genotype)
unique(teosinte_genotype$Group)
```

Transposing teosinte and maize genotype data
```{r}
transposed_teosinte_genotype <- t(teosinte_genotype)
nrow(transposed_teosinte_genotype)
ncol(transposed_teosinte_genotype)

transposed_maize_genotype <- t(maize_genotype)
nrow(transposed_maize_genotype)
ncol(transposed_maize_genotype)

#show the head of the first column for transposed_maize_genotype
head(transposed_maize_genotype[,1])
#remove the first three rows of tranposed_genotype as they are not important for merging
transposed_maize_genotype_without_header <- transposed_maize_genotype[-(1:3),]
#show the head of the new transposed_maize_genotype_without_header (without the first three rows of tranposed_maize_genotype)
head (transposed_maize_genotype_without_header[,1])

head(transposed_teosinte_genotype[,1])
transposed_teosinte_genotype_without_header<-transposed_teosinte_genotype[-(1:3),]
head(transposed_teosinte_genotype_without_header[,1])




```
Merge tranposed_maize_genotype_without_header with snp_position.txt
```{r}
transposed_maize_genotype_without_header<-cbind(colnames(genotype_data[4:ncol(genotype_data)]),transposed_maize_genotype_without_header)

colnames(transposed_maize_genotype_without_header)[1]<-paste("SNP_ID")

maize_genotype_snp<-merge(snp_position,transposed_maize_genotype_without_header, by.x="SNP_ID", by.y="SNP_ID")
```

Remove unwanted columns from maize_genotype_snp. We only want the SNP_ID, the chromosome, the position and the genotype data.(Columns 1,3,4,16 to end)
```{r}
#colnames(maize_genotype_snp[1:20])
#colnames (maize_genotype_snp[c(1:20)])
#colnames(maize_genotype_snp[c(1,3,4,16:ncol(maize_genotype_snp))])
maize_genotype_snp<-maize_genotype_snp[c(1,3,4,16:ncol(maize_genotype_snp))]

#maize_genotype_snp[,-c(1:3)]<lapply(maize_genotype_snp[,-c(1:3)],as.character)
```
Merge transposed_teosinte_genotype_without_header with snp_position.txt
```{r}
transposed_teosinte_genotype_without_header<-cbind(colnames(genotype_data[4:ncol(genotype_data)]),transposed_teosinte_genotype_without_header)

colnames(transposed_teosinte_genotype_without_header)[1]<-paste("SNP_ID")

teosinte_genotype_snp<-merge(snp_position,transposed_teosinte_genotype_without_header, by.x="SNP_ID", by.y="SNP_ID")
```
Remove unwanted columns from teosinte_genotype_snp.
```{r}
teosinte_genotype_snp<-teosinte_genotype_snp[c(1,3,4,16:ncol(teosinte_genotype_snp))]

#sets columns 4:ncol(teosinte_genotype_snp) to character value
#teosinte_genotype_snp[,-c(1:3)]<-lapply(teosinte_genotype_snp[,-c(1:3)], as.character)

```

Seperate maize_genotype_snp into individual chromosome files.
```{r}
unique(maize_genotype_snp[,2])
for(i in 1:10){
      temp_chromosome_df<-maize_genotype_snp[maize_genotype_snp[,2]==i,]
      
      ordered_temp_chromosome_df<- temp_chromosome_df[order(as.numeric(temp_chromosome_df$Position)),]
      
      write.csv(ordered_temp_chromosome_df, sprintf("maize_chromosome_%d.csv",i), row.names=FALSE)

}
testing<-read.csv("maize_chromosome_1.csv")
```

Sort maize_chromosome_%d by position (descending) with ?/? substituted for -/-
```{r}
for(i in 1:10){
      temp_chromosome_df<-maize_genotype_snp[maize_genotype_snp[,2]==i,]
      temp_chromosome_df[,-c(1:3)]<-lapply(temp_chromosome_df[,-c(1:3)],as.character)
     
      #replace "?/?" with "-/-"
      temp_chromosome_df[temp_chromosome_df=="?/?"] <-"-/-"
      
      descending_order_temp_chromosome_df<- temp_chromosome_df[order(-(as.numeric(temp_chromosome_df$Position))),]
      
      write.csv(descending_order_temp_chromosome_df, sprintf("descending_order_maize_chromosome_%d.csv",i), row.names=FALSE)
}
descending_order_test<-read.csv("descending_order_maize_chromosome_1.csv")


```

Seperate teosinte_genotype_snp into individual chromosome files
```{r}

for(i in 1:10){
      temp_t_chromosome_df<-teosinte_genotype_snp[teosinte_genotype_snp[,2]==i,]
      
      ordered_temp_t_chromosome_df<- temp_t_chromosome_df[order(as.numeric(temp_t_chromosome_df$Position)),]
      
      write.csv(ordered_temp_t_chromosome_df, sprintf("teosinte_chromosome_%d.csv",i), row.names=FALSE)
      }

```
Sort teosinte_chromosome%d by position (descending) with ?/? substituted for -/-
```{r}
for(i in 1:10){
  
      temp_teosinte_chromosome_df<-teosinte_genotype_snp[teosinte_genotype_snp[,2]==i,]
      temp_teosinte_chromosome_df[,-c(1:3)]<-lapply(temp_teosinte_chromosome_df[,-c(1:3)],as.character)
      #replace "?/?" with "-/-"
      temp_teosinte_chromosome_df[temp_teosinte_chromosome_df=="?/?"] <-"-/-"
      
      descending_order_temp_teosinte_chromosome_df<- temp_teosinte_chromosome_df[order(-(as.numeric(temp_teosinte_chromosome_df$Position))),]
      
      write.csv(descending_order_temp_teosinte_chromosome_df, sprintf("descending_order_teosinte_chromosome_%d.csv",i), row.names=FALSE)
}
descending_order_teosinte_test<-read.csv("descending_order_teosinte_chromosome_1.csv")


```
Import the reshape2, dplyr and ggplot2 packages.
```{r}
install.packages("reshape2")
install.packages("dplyr")
install.packages("ggplot2")

library(ggplot2)
library(dplyr)
library(reshape2)

```
Plot the total number of SNPs in out dataset on each chromosome.
```{r}
#typeof(genotype_data)
transposed_fang<-t(genotype_data)
#unique(genotype_data$Group)

transposed_fang<-as.data.frame(transposed_fang,row.names = FALSE)
groups<-transposed_fang[3,]
#typeof(groups)
transposed_fang<-transposed_fang[-(1:3),]

transposed_fang<-cbind(colnames(genotype_data[4:ncol(genotype_data)]),transposed_fang)#this adds the headers from the untransposed data back into a column form
colnames(transposed_fang)[1]<-paste("SNP_ID")
transposed_genotype_snp<-merge(snp_position,transposed_fang, by.x="SNP_ID", by.y="SNP_ID")
dim(transposed_genotype_snp)
dim(groups)
transposed_genotype_snp<-transposed_genotype_snp[c(1,3,4,16:ncol(transposed_genotype_snp))]
dim(transposed_genotype_snp)
groups<-cbind(0,0,0,groups)
colnames(groups)<-colnames(transposed_genotype_snp)
transposed_genotype_snp<-rbind(groups,transposed_genotype_snp)
groups_unique<-unique(genotype_data$Groups)
snp_chromosome_count_df<-data.frame()
for (i in 1:10){
  temp_chromosome_df<-transposed_genotype_snp[transposed_genotype_snp[,2]==i,]
  chromosome_count<-nrow(temp_chromosome_df)
  chromosome_count<-cbind(i,chromosome_count)
  snp_chromosome_count_df<-rbind(snp_chromosome_count_df, chromosome_count)
}
colnames(snp_chromosome_count_df)
colnames(snp_chromosome_count_df)<-c("Chromosome","Number_of_SNPs")
  colnames(snp_chromosome_count_df)

ggplot(snp_chromosome_count_df) + geom_point(aes(x=Chromosome, y=Number_of_SNPs))+ ggtitle("Total number of SNPs on each chromosome")

```
How did groups contribute to the number of SNPs?
```{r}
 transposed_genotype_snp[,-c(1:3)]<-lapply(transposed_genotype_snp[,-c(1:3)],as.character)


SNPs_per_group<-data.frame()
groups_unique<-unique(genotype_data$Group)
for (i in 1:length(groups_unique)){
 temp_group_df<-transposed_genotype_snp[,transposed_genotype_snp[1,]== groups_unique[i]]
  
   #temp_group_df<-data.frame(transposed_genotype_snp[,transposed_genotype_snp[1,]== groups_unique[i]], stringsAsFactors = FALSE)
   #temp_group_df[,-c(1:3)]<-lapply(temp_group_df[,-c(1:3)],as.character)
  
  group_count<-ncol(temp_group_df)
  group_count<-cbind(groups_unique[i],group_count)
  SNPs_per_group<-rbind(SNPs_per_group, group_count)
}
colnames(SNPs_per_group)
colnames(SNPs_per_group)<-c("Group", "Number_of_SNPs")
SNPs_per_group
SNPs_per_group$Number_of_SNPs<-as.numeric(levels(SNPs_per_group$Number_of_SNPs))[SNPs_per_group$Number_of_SNPs]
ggplot(SNPs_per_group, aes(x=Group, y=Number_of_SNPs))+geom_point()+theme(axis.text.x=element_text(angle=90)) +ggtitle("Number of SNPs by Group")

```
Missing Data and amount of heterozogosity and Sort by group and then by sample_ID
```{r}
genotype_data<-as.data.frame(genotype_data)
homozygous<-as.data.frame(matrix(ncol=1,nrow=nrow(genotype_data))) #adds homozygous column
total_NA<-as.data.frame(matrix(ncol=1,nrow=nrow(genotype_data)))
heterozygous<-as.data.frame(matrix(ncol=1,nrow=nrow(genotype_data)))
homozygous_df<-cbind(genotype_data[c(1:3)],homozygous,heterozygous,total_NA,genotype_data[c(4:ncol(genotype_data))])
colnames(homozygous_df)[4]<-paste("Homozygous_percent")
colnames(homozygous_df)[5]<-paste("Heterozygous_percent")
colnames(homozygous_df)[6]<-paste("NA_percent")
homozygous_df[homozygous_df=="?/?"] <- NA
for (i in 1:nrow(homozygous_df))
{
  SNPs<-homozygous_df[i,c(7:ncol(homozygous_df))]
  counter<-0
  
  for(j in 1:ncol(SNPs)){
    if (is.na(SNPs[1,j])==FALSE){
   if((SNPs[1,j] == "A/A") | (SNPs[1,j] == "T/T")| (SNPs[1,j] == "C/C") | (SNPs[1,j] == "G/G")){
     counter<-counter +1;
   }
   }
  }
  homozygous_df[i,6]<-sum(is.na(SNPs))/ncol(SNPs)
  homozygous_df[i,4]<-counter/ncol(SNPs)
  homozygous_df[i,5]<-1-(homozygous_df[i,4])-(homozygous_df[i,6])
}
homozygous_df_sorted<-homozygous_df[order(homozygous_df$Group),]
homozygous_df_sorted<-homozygous_df_sorted[order(homozygous_df_sorted$Sample_ID),]
```

Graph homozygousity, heterozygosity and NA_percentage
```{r}

temp_df<- melt(homozygous_df_sorted[c(1:6)], id= c("Sample_ID", "JG_OTU", "Group"))

ggplot(temp_df, aes(x=temp_df$Sample_ID, y=temp_df$value, fill=temp_df$variable))+geom_bar(stat="identity")+ xlab("Sample_Id") + ylab("Proportion")+ ggtitle("Proportionality of homozygous,heterozygous and NA SNPs")+ theme(legend.title=element_blank())
```


Homozygous/heterozygous/NA SNPs for maize
```{r}
maize_genotype<-as.data.frame(maize_genotype)
homozygous<-as.data.frame(matrix(ncol=1,nrow=nrow(maize_genotype))) #adds homozygous column
total_NA<-as.data.frame(matrix(ncol=1,nrow=nrow(maize_genotype)))
heterozygous<-as.data.frame(matrix(ncol=1,nrow=nrow(maize_genotype)))
homozygous_df<-cbind(maize_genotype[c(1:3)],homozygous,heterozygous,total_NA,maize_genotype[c(4:ncol(maize_genotype))])
colnames(homozygous_df)[4]<-paste("Homozygous_percent")
colnames(homozygous_df)[5]<-paste("Heterozygous_percent")
colnames(homozygous_df)[6]<-paste("NA_percent")
homozygous_df[homozygous_df=="?/?"] <- NA
for (i in 1:nrow(homozygous_df))
{
  SNPs<-homozygous_df[i,c(7:ncol(homozygous_df))]
  counter<-0
  
  for(j in 1:ncol(SNPs)){
    if (is.na(SNPs[1,j])==FALSE){
   if((SNPs[1,j] == "A/A") | (SNPs[1,j] == "T/T")| (SNPs[1,j] == "C/C") | (SNPs[1,j] == "G/G")){
     counter<-counter +1;
   }
   }
  }
  homozygous_df[i,6]<-sum(is.na(SNPs))/ncol(SNPs)
  homozygous_df[i,4]<-counter/ncol(SNPs)
  homozygous_df[i,5]<-1-(homozygous_df[i,4])-(homozygous_df[i,6])
}
homozygous_df_sorted<-homozygous_df[order(homozygous_df$Group),]
homozygous_df_sorted<-homozygous_df_sorted[order(homozygous_df_sorted$Sample_ID),]

temp_df<- melt(homozygous_df_sorted[c(1:6)], id= c("Sample_ID", "JG_OTU", "Group"))

ggplot(temp_df, aes(x=temp_df$Sample_ID, y=temp_df$value, fill=temp_df$variable))+geom_bar(stat="identity")+ xlab("Sample_Id") + ylab("Proportion")+ ggtitle("Proportionality of homozygous,heterozygous and NA SNPs in Maize")+ theme(legend.title=element_blank())
```
Homozygous/Heterozygous/NA SNPs in teosinte
```{r}
teosinte_genotype<-as.data.frame(teosinte_genotype)
homozygous<-as.data.frame(matrix(ncol=1,nrow=nrow(teosinte_genotype))) #adds homozygous column
total_NA<-as.data.frame(matrix(ncol=1,nrow=nrow(teosinte_genotype)))
heterozygous<-as.data.frame(matrix(ncol=1,nrow=nrow(teosinte_genotype)))
homozygous_df<-cbind(teosinte_genotype[c(1:3)],homozygous,heterozygous,total_NA,teosinte_genotype[c(4:ncol(teosinte_genotype))])
colnames(homozygous_df)[4]<-paste("Homozygous_percent")
colnames(homozygous_df)[5]<-paste("Heterozygous_percent")
colnames(homozygous_df)[6]<-paste("NA_percent")
homozygous_df[homozygous_df=="?/?"] <- NA
for (i in 1:nrow(homozygous_df))
{
  SNPs<-homozygous_df[i,c(7:ncol(homozygous_df))]
  counter<-0
  
  for(j in 1:ncol(SNPs)){
    if (is.na(SNPs[1,j])==FALSE){
   if((SNPs[1,j] == "A/A") | (SNPs[1,j] == "T/T")| (SNPs[1,j] == "C/C") | (SNPs[1,j] == "G/G")){
     counter<-counter +1;
   }
   }
  }
  homozygous_df[i,6]<-sum(is.na(SNPs))/ncol(SNPs)
  homozygous_df[i,4]<-counter/ncol(SNPs)
  homozygous_df[i,5]<-1-(homozygous_df[i,4])-(homozygous_df[i,6])
}
homozygous_df_sorted<-homozygous_df[order(homozygous_df$Group),]
homozygous_df_sorted<-homozygous_df_sorted[order(homozygous_df_sorted$Sample_ID),]

temp_df<- melt(homozygous_df_sorted[c(1:6)], id= c("Sample_ID", "JG_OTU", "Group"))

ggplot(temp_df, aes(x=temp_df$Sample_ID, y=temp_df$value, fill=temp_df$variable))+geom_bar(stat="identity")+ xlab("Sample_Id") + ylab("Proportion")+ ggtitle("Proportionality of homozygous,heterozygous and NA SNPs in Teosinte")+ theme(legend.title=element_blank())

```

